sdks/python/apache_beam/runners/worker/sdk_worker.py

```diff
@@ -148,3 +152,87 @@ def process_bundle(self, request, instruction_id):
   def process_bundle_progress(self, request, instruction_id):
     # It is an error to get progress for a not-in-flight bundle.
     return self.bundle_processors.get(instruction_id).metrics()
+
+
+class GrpcStateHandler(object):
+
+  _DONE = object()
+
+  def __init__(self, state_stub):
+    self._lock = threading.Lock()
+    self._state_stub = state_stub
+    self._requests = queue.Queue()
+    self._responses = queue.Queue()
+    self._responses_by_id = {}
+    self._last_id = 0
+    self._exc_info = None
+
+  def start(self):
+    self._done = False
+
+    def request_iter():
+      while True:
+        request = self._requests.get()
+        if request is self._DONE or self._done:
+          break
+        yield request
+    responses = self._state_stub.State(request_iter())
+
+    def pull_responses():
+      try:
+        for response in responses:
+          self._responses.put(response)
+          if self._done:
+            break
+      except:  # pylint: disable=bare-except
+        self._exc_info = sys.exc_info()
+        raise
+    reader = threading.Thread(target=pull_responses, name='read_state')
+    reader.daemon = True
+    reader.start()
+
+  def done(self):
+    self._done = True
+    self._requests.put(self._DONE)
+
+  def blocking_get(self, state_key, instruction_reference):
+    response = self._blocking_request(
+        beam_fn_api_pb2.StateRequest(
+            instruction_reference=instruction_reference,
+            state_key=state_key,
+            get=beam_fn_api_pb2.StateGetRequest()))
+    if response.get.continuation_token:
+      raise NotImplementedErrror
+    return response.get.data
+
+  def blocking_append(self, state_key, data, instruction_reference):
+    self._blocking_request(
+        beam_fn_api_pb2.StateRequest(
+            instruction_reference=instruction_reference,
+            state_key=state_key,
+            append=beam_fn_api_pb2.StateAppendRequest(data=data)))
+
+  def blocking_clear(self, state_key, instruction_reference):
+    self._blocking_request(
+        beam_fn_api_pb2.StateRequest(
+            instruction_reference=instruction_reference,
+            state_key=state_key,
+            clear=beam_fn_api_pb2.StateClearRequest()))
+
+  def _blocking_request(self, request):
+    request.id = self._next_id()
+    self._requests.put(request)
+    while request.id not in self._responses_by_id:
+      with self._lock:
+        if request.id not in self._responses_by_id:
+          response = self._responses.get(timeout=2)
```

Comment:
<!--new_thread; commit:3dfe862220afbc5e57d7ac674a329cb5e66164cf; resolved:0-->
Why `timeout=2`? Could not this raise an empty queue exception?

Reply:
No reason to have 2, changed to 1. 

Basically, this should periodically check for exceptions, and periodically release the lock in case the data for another thread came in.

However, this code isn't at all the easiest to follow. Changed to do the demultiplexing in the reader loop and use events to block. (In addition to being simpler, this should better extend itself to being able to return future-like objects for side inputs and state in the future.)